text/0016-generalized-entity-store.md

```diff
@@ -0,0 +1,113 @@
+
+- Implementation PR(s): https://github.com/cedar-policy/cedar/pull/171, which is a pull request from https://github.com/prakol16/cedar
+
+## Timeline
+
+- Start Date: 2023-07-11
+- Date Entered FCP:
+- Date Accepted:
+- Date Landed:
+
+## Summary
+
+Entity storage ideally should not be restricted to JSON files. We should either allow users to add their own custom storage methods or add support for some non-JSON common storage methods (e.g. databases) ourselves.
+
+## Basic example
+
+
+```rust
+struct MyOwnDatabase {
+    // ...
+}
+
+impl EntityDatabase for MyOwnDatabase {
+    fn get_entity_of_uid(&self, uid: &EntityUID) -> Option<Entity> {
+        // custom implementation which could e.g. make a request to a database
+    }
+}
+```
+
+This could be code written by users, or we could hide `EntityDatabase` and instead, there could be several storage solutions (other than the default JSON one) that we write ourselves.
+
+## Motivation
+
+  - Currently, every request reads the entire JSON file, reparsing all of the attributes, even though the overwhelming majority of requests will need to read only a few entries of the entity store. In general, with many entities, it may not even be feasible to load all of the entities into memory at once.
+
+  - In the future, we can turn partial expression evaluations into queries so that we can efficiently find e.g. all of the photos that a user has access to. Ideally, we would leverage existing query languages such as SQL and existing storage solutions which are already optimized to respond to these kinds of queries efficiently, rather than try to rebuild this on our own.
+
+  - Some users may already have infrastructure that uses databases or other storage methods for their applications. This makes integrating Cedar into existing software easier, because users are not boxed in to one particular storage format.
+
+
+## Detailed design
+
+
+We would generalize the authorizer to accept any `EntityDatabase`, not just `Entities`:
+
+```rust
+impl Authorizer {
+    // old:
+    pub fn is_authorized_old(&self, q: &Request, pset: &PolicySet, entities: &Entities) -> Response
+
+    // proposed change:
+    pub fn is_authorized<T : EntityDatabase>(&self, q: &Request, pset: &PolicySet, entities: &T) -> Response
+}
+```
+
+Here is the current interface for `EntityDatabase` (see [the source](https://github.com/prakol16/cedar/blob/e5df52e118e378c2142aaa6d9a4781f0f67b8064/cedar-policy-core/src/entities.rs#L265) for more comments and the implementation of the trait for `Entities`):
+
+```rust
+pub trait EntityDatabase {
+    /// Given a uid, get the corresponding entity if it exists
+    /// The entity should already have an `ancestors` attribute which represents
+    /// the transitive closure ancestors, not the direct parents
+    fn get_entity_of_uid(&self, uid: &EntityUID) -> Option<Entity>;
+
+    /// Returns whether this database should return expressions for unknown entities
+    fn is_partial(&self) -> bool;
```

Comment:
What does this mean?

Reply:
It's the same as `Entities::mode`. The `EntityDatabase` is basically an abstraction of `Entities` but with the hash map replaced by a function, so that is where this comes from.

Reply:
Thank you! Would it make sense to replicate `Entities::mode` rather than introduce a new concept? The thing that I tripped over reading this comment as an uneducated person was "whether this database should return expressions for unknown entities", for a couple of reasons:

1. it seems to be returning whether the storage _will_ return expressions rather than a configuration of the storage to tell it how it should behave? (so me just being a pedantic person reading the comment, sorry, honestly I am just trying to understand).
2. where would the expressions be returned? what would they look like? It looks like [`Entities::entity(uid)` will return an `Expr::unknown` with the uid when it's in `Mode::Partial`](https://github.com/cedar-policy/cedar/blob/d9201161d8ce45e015c8bcb1cd4191c11039709a/cedar-policy-core/src/entities.rs#L85). Having dug in (and thank you for the link!) it makes more sense now. Would it make sense to add a couple of lines to the comment to clarify, maybe something like:

```suggestion
    /// Returns whether this data store is complete (`Mode::Concrete`) and failed dereferences are an error
    /// (`entity()` will return a `Dereference::NoSuchEntity` on failure) or whether this data store is partial 
    /// (`Mode::Partial`) and failed dereferences will result in a residual (`entity()` will return a `Dereference::Residual`
    /// with an expression representing the entity).
    fn mode(&self) -> Mode;
```

(clearly I'm super-new here so that suggestion is likely wrong in multiple ways, but hopefully it gets the idea across?)

Reply:
Yeah, this makes sense. The implementation is still a rough draft so the comments are not really full explanations. Your comments seem like reasonable clarifications.

The reason it returns a boolean and doesn't reuse `Entities::mode` is simply because `Entities::mode` was not a public struct before, and so this path had slightly less resistance -- I'll probably go back and update it to use `Entities::mode` and make it public.

By the way, I'm also super new (I just started looking at cedar last week) so I think we're in the same boat here :)